### PR TITLE
Self Transfer Icon, Load Transactions on Scroll

### PR DESCRIPTION
**Pull request by sircoderin**

### CODE REVIEW

File: popup.html [hunk 1]
```diff
@@ -131,32 +131,7 @@
...
-                <div id="tx4" class="transaction">
-                    <span id="tx-type-4" class="tx-type"></span>
-                    <span id="tx-info-4" class="tx-info"></span>
-                </div>
-                <div id="tx5" class="transaction">
-                    <span id="tx-type-5" class="tx-type"></span>
-                    <span id="tx-info-5" class="tx-info"></span>
-                </div>
-            </div>
+            <div id = "transactions-container" class="transactions-container" style="overflow:scroll; height:259px;"> </div>
```
Comment (andrewtookay) [inline popup.html]: remove the class and move the styling of the element into the CSS file under #transactions-container

File: popup.js [hunk 2]
```diff
@@ -44,10 +44,15 @@ function refreshBalance(value) {
 }
 
 function processLatestTransactions(latestTransactions = []) {
+
     if (latestTransactions.length >= 1) {
-        for (let i = latestTransactions.length - 1; i >= 0; i--) {
-            txInfo = document.getElementById("tx-info-" + (latestTransactions.length - 1 - i));
-            txType = document.getElementById("tx-type-" + (latestTransactions.length - 1 - i));
+        let divNumber = 0;
+        //TODO add date-time to the transaction to be more precise about the ordering ?
```
Comment (andrewtookay) [inline popup.js]: change the comment to TODO_ADOT_LOW for now

File: popup.js [hunk 3]
```diff
@@ -57,57 +62,103 @@ function processLatestTransactions(latestTransactions = []) {
 
             txInfo.innerHTML += "<br>";
 
-            if (latestTransactions[i].address !== undefined) {
-                txInfo.innerHTML += latestTransactions[i].address;
-            } else if (latestTransactions[i].category == "send") {
-                if (latestTransactions[i].amount === -0.1 && latestTransactions[i].fee === -0.1) {
+            const currentTransactionAddress = latestTransactions[i].address;
```
Comment (andrewtookay) [inline popup.js]: maybe set all const variables at the start of the for loop?

File: popup.js [hunk 4]
```diff
@@ -57,57 +62,103 @@ function processLatestTransactions(latestTransactions = []) {
...
+                if (currentTransactionAmount === -0.1 && latestTransactions[i].fee === -0.1) {
                     txInfo.innerHTML += "BlockchainDNS Registration";
-                    
+
                     txType.innerHTML = "REG";
                     txType.style.color = "rgb(30, 118, 210)";
 
                     continue;
-                } else if (latestTransactions[i].amount === -0.005 && latestTransactions[i].fee === -0.005) {
+                } else if (currentTransactionAmount === -0.005 && latestTransactions[i].fee === -0.005) {
```
Comment (andrewtookay) [inline popup.js]: the fee is still used straight from the tx array

File: popup.js [hunk 5]
```diff
@@ -57,57 +62,103 @@ function processLatestTransactions(latestTransactions = []) {
...
+        const currentTransactionCategory = latestTransactions[transactionIndex].category;
+        return currentTransactionCategory == "generate" || currentTransactionCategory == "receive";
+    }
+
+    function isToSelf(transactionIndex, currentTransactionAddress) {
+        return transactionIndex >= 1 && isGenerateOrReceive(transactionIndex - 1) &&
+            currentTransactionAddress == latestTransactions[transactionIndex - 1].address;
+    }
+
+    function createDivForTransaction(index) {
```
Comment (andrewtookay) [inline popup.js]: maybe this would be better like createTxContainerDiv?

File: popup.js [hunk 6]
```diff
@@ -57,57 +62,103 @@ function processLatestTransactions(latestTransactions = []) {
...
+        div.id = "th" + index;
+        div.className = "transaction";
+
+        div.appendChild(createGenericDiv(index, "tx-type"));
+        div.appendChild(createGenericDiv(index, "tx-info"));
+
+        document.getElementById('transactions-container').appendChild(div);
+    }
+
+    function createGenericDiv(i, tx_data) {
```
Comment (andrewtookay) [inline popup.js]: and createTxInnerDiv?

File: popup.js [hunk 7]
```diff
@@ -57,57 +62,103 @@ function processLatestTransactions(latestTransactions = []) {
...
+        return transactionIndex >= 1 && isGenerateOrReceive(transactionIndex - 1) &&
+            currentTransactionAddress == latestTransactions[transactionIndex - 1].address;
+    }
+
+    function createDivForTransaction(index) {
+        if (document.getElementById("th" + index) != null)
+            return;
+
+        var div = document.createElement('div');
+        div.id = "th" + index;
```
Comment (andrewtookay) [inline popup.js]: use tx-x format, similar to tx-info-x and tx-type-x

File: popup.js [hunk 8]
```diff
@@ -57,57 +62,103 @@ function processLatestTransactions(latestTransactions = []) {
...
+                    txType.innerHTML = "ERR";
+                    txType.style.color = "rgb(255, 0, 0)";
             }
         }
     } else {
         document.getElementById("tx-info-0").innerHTML = "&emsp;&emsp;&emsp;&emsp;No transaction history.";
     }
+
+    function isGenerateOrReceive(transactionIndex) {
+        const currentTransactionCategory = latestTransactions[transactionIndex].category;
```
Comment (andrewtookay) [inline popup.js]: do we really need this const and even the function? aren't self transfers pairs of send-receive? send-generate souldn't really apply

File: popup.js [hunk 9]
```diff
@@ -57,57 +62,103 @@ function processLatestTransactions(latestTransactions = []) {
...
+        return currentTransactionCategory == "generate" || currentTransactionCategory == "receive";
+    }
+
+    function isToSelf(transactionIndex, currentTransactionAddress) {
+        return transactionIndex >= 1 && isGenerateOrReceive(transactionIndex - 1) &&
+            currentTransactionAddress == latestTransactions[transactionIndex - 1].address;
```
Comment (andrewtookay) [inline popup.js]: add a check for amount as well?

**Review by andrewtookay — CHANGES_REQUESTED** [review]

File: popup.html (hunk 1 above)
Comment (sircoderin) [reply popup.html]: Why remove the class though and not move the style under .transactions-container?

Comment (andrewtookay) [reply popup.html]: well, we won't really reuse this class as there's going to be only one container for all transactions and we already use the id to get the element

File: popup.js [hunk 10]
```diff
@@ -12,26 +12,29 @@ var state = {
     rpcPort: "31050",
     ready: 0, // ready on 8
     readyDOM: false,
-    useDebug: false
+    useDebug: false,
+    transactionsReturned: 20
 }
 
+var scrollEventListenerAdded = false
```
Comment (andrewtookay) [inline popup.js]: add this to the state, I don't think we should do lone global variables, maybe rename it to txScrollEventListenerAdded?

File: popup.js [hunk 11]
```diff
@@ -44,70 +47,136 @@ function refreshBalance(value) {
...
+                        txType.innerHTML = "&#8651;";
+                        txType.style.color = "rgb(138,43,226)";
+                        i--;
+                    } else {
+                        txType.innerHTML = "&#8594;";
+                        txType.style.color = "rgb(255, 0, 0)";
+                    }
+                    break;
+                case "immature":
+                    txType.innerHTML = "IMM";
```
Comment (andrewtookay) [inline popup.js]: due to the increased font size this appears too big and overlaps with the tx info

File: popup.css [hunk 12]
```diff
@@ -678,3 +671,12 @@ input:checked+.slider:before {
 #notification.info {
     background-color: rgb(54, 121, 197);
 }
+
+#transactions-container {
+    width: 340px;
+    height:259px;
```
Comment (andrewtookay) [inline popup.css]: 259px seems to be more than the actual height of the container

File: popup.js [hunk 13]
```diff
@@ -44,70 +47,136 @@ function refreshBalance(value) {
...
+        return tx;
+    }
+
+    function addScrollListener() {
+        document.getElementById("transactions-container")
+            .addEventListener('scroll', function (event) {
+                var element = event.target;
+                if (element.scrollHeight - element.scrollTop === element.clientHeight) {
+                    state.transactionsReturned = state.transactionsReturned + 10;
+                    setTimeout(loopRefreshWallet, 10);
```
Comment (andrewtookay) [inline popup.js]: this starts a refresh loop instead of doing one refresh

**Review by andrewtookay — CHANGES_REQUESTED** [review]
also remove the .DS_Store file

File: popup.css [hunk 14]
```diff
@@ -678,5 +677,9 @@ input:checked+.slider:before {
     flex-flow: column;
     align-items: center;
     margin-top: auto;
-    overflow:scroll; 
+    overflow-x: hidden;
+    overflow-y: auto;
 }
+/* #transactions-container::-webkit-scrollbar {
```
Comment (andrewtookay) [inline popup.css]: uncomment this as the scrollbar still appears

File: popup.js [hunk 15]
```diff
@@ -13,11 +13,9 @@ var state = {
     ready: 0, // ready on 8
     readyDOM: false,
     useDebug: false,
-    transactionsReturned: 20
+    transactionsReturned: 20
```
Comment (andrewtookay) [inline popup.js]: increase the initial set of transactions to 60 and then add 30 each time you scroll to the bottom

**Review by andrewtookay — CHANGES_REQUESTED** [review]
remove the .DS_Store file as it is still present

File: popup.js (hunk 15 above)
Comment (andrewtookay) [reply popup.js]: when the wallet is still synchronizing the requests are a bit laggy and the scrolling and loading of transactions doesn't work smoothly, maybe a loading circle can be added in the middle of the container

File: popup.js [hunk 16]
```diff
@@ -43,71 +46,133 @@ function refreshBalance(value) {
     }
 }
 
+function addScrollListener() {
+    var container = document.getElementById("transactions-container");
+    container.addEventListener('scroll', function (event) {
+        var element = event.target;
+        if (element.scrollHeight - element.scrollTop === element.clientHeight) {
+            state.transactionsReturned = state.transactionsReturned + 30;
+            refreshWallet();
```
Comment (andrewtookay) [inline popup.js:60]: refreshWallet also refreshes the balance and that's not needed here

File: popup.css (hunk 12 above)
Comment (andrewtookay) [reply popup.css]: this should be 222px as that's 37 * 6 tx, the container is too close to the execute operation button because of this

**Review by andrewtookay — CHANGES_REQUESTED** [review]

File: popup.js [hunk 17]
```diff
@@ -162,6 +152,20 @@ function processLatestTransactions(latestTransactions = []) {
...
+    function updateTx(txInfo, txType, txTypeText, txTypeColor, txTypeFontSize, txInfoText) {
+        txInfo.innerHTML += txInfoText;
+
+        txType.innerHTML = txTypeText;
+        txType.style.color = txTypeColor;
+        txType.style.fontSize = txTypeFontSize;
+    }
+
+    function getMiningIconLink(color, width, height) {
+        return "<img src='https://api.iconify.design/mdi/pickaxe.svg?color=" +
```
Comment (andrewtookay) [inline popup.js]: add the code of the SVG here instead of loading it from an external source

File: popup.js [hunk 18]
```diff
@@ -13,7 +13,8 @@ var state = {
     ready: 0, // ready on 8
     readyDOM: false,
     useDebug: false,
-    transactionsReturned: 60
+    transactionsReturnedInitially: 60,
```
Comment (andrewtookay) [inline popup.js]: this is more of a total as this number increases during usage, just transactionsReturned is a good name

File: popup.js [hunk 19]
```diff
@@ -13,7 +13,8 @@ var state = {
     ready: 0, // ready on 8
     readyDOM: false,
     useDebug: false,
-    transactionsReturned: 60
+    transactionsReturnedInitially: 60,
+    transactionReturnedOnScrollDown: 30
```
Comment (andrewtookay) [inline popup.js]: maybe this should be transactionsAddedOnScroll?

File: popup.js [hunk 20]
```diff
@@ -51,18 +52,19 @@ function addScrollListener() {
...
 }
 
+
 function processLatestTransactions(latestTransactions = []) {
 
     if (latestTransactions.length >= 1) {
         let divNumber = 0;
         //TODO_ADOT_LOW add date-time to the transaction to be more precise about the ordering ?
-        for (let i = latestTransactions.length - 1; i >= 0; i--, divNumber++) {
+        for (var i = latestTransactions.length - 1; i >= 0; i--, divNumber++) {
```
Comment (andrewtookay) [inline popup.js]: var might cause problems here

File: popup.js [hunk 21]
```diff
@@ -86,48 +88,36 @@ function processLatestTransactions(latestTransactions = []) {
...
+                    updateTx(txInfo, txType, "UPD", "rgb(30, 118, 210)", "16px", "BlockchainDNS Update");                    continue;
                 }
             }
 
             switch (category) {
-                case "generate": case "receive":
-                    txType.innerHTML = "&#8592;";
-                    txType.style.color = "rgb(50, 205, 50)";
+                case "generate":
+                    updateTx(txInfo, txType, getMiningIconLink("green", 22, 22));
```
Comment (andrewtookay) [inline popup.js]: these updateTx don't have a txInfoText sent as parameter and that causes undefined to appear

File: popup.js [hunk 22]
```diff
@@ -162,6 +152,20 @@ function processLatestTransactions(latestTransactions = []) {
         tx.className = tx_data;
         return tx;
     }
+
+    function updateTx(txInfo, txType, txTypeText, txTypeColor, txTypeFontSize, txInfoText) {
```
Comment (andrewtookay) [inline popup.js]: txTypeText should be txTypeContent as it's not only text

File: popup.js [hunk 23]
```diff
@@ -322,16 +333,24 @@ function hideNotification() {
...
         sendCommand(url, "getbalance", [], refreshBalance, (reqStatus, errMessage) => {
             processRequestFail(state.useDebug, reqStatus, errMessage, "refreshWallet getbalance");
         }, state.useDebug);
-        sendCommand(url, "listtransactions", ["*", state.transactionsReturned, 0], processLatestTransactions, (reqStatus, errMessage) => {
+        sendCommand(url, "listtransactions", ["*", state.transactionsReturnedInitially, 0], processLatestTransactions, (reqStatus, errMessage) => {
             processRequestFail(state.useDebug, reqStatus, errMessage, "refreshWallet listtransactions");
         }, state.useDebug);
+
+        await sleep(3000);
+        document.getElementById('load').style.display = "none";
```
Comment (andrewtookay) [inline popup.js]: this should happen in the success/fail methods that follow up "listtransactions"

**Review by andrewtookay — CHANGES_REQUESTED** [review]

**Review by andrewtookay — APPROVED** [review]

**Pull request merged**